examples/mlperf/initializers.py

```diff
@@ -67,4 +65,12 @@ def __call__(self, x:Tensor):
     assert self.normalized_shape == x.shape[-len(self.normalized_shape):], f"last dimensions of {x.shape} must match {self.normalized_shape}"
     xn = x.cast(dtypes.float32).layernorm(eps=self.eps, axis=self.axis).cast(x.dtype)
     if not self.elementwise_affine: return xn
-    return (xn * self.weight.cast(dtypes.default_float) + self.bias.cast(dtypes.default_float))
\ No newline at end of file
+    return (xn * self.weight.cast(dtypes.default_float) + self.bias.cast(dtypes.default_float))
+
+def scaled_dot_product_attention_bert(query:Tensor, key:Tensor, value:Tensor, attn_mask:Tensor=None, dropout_p:float=0.0, is_causal:bool=False):
+  assert all_int(query.shape), f"does not support symbolic shape {query.shape}"
+  if is_causal: attn_mask = Tensor.ones(query.shape[-2], key.shape[-2], requires_grad=False, device=query.device).tril(0).cast(dtypes.bool)
+  if attn_mask is not None and attn_mask.dtype == dtypes.bool: attn_mask = (attn_mask == 0).where(-float("inf"), 0)
+  # NOTE: During training we get -inf from key @ query, so we cast to float
+  qk = query.cast(dtypes.float32) @ key.cast(dtypes.float32).transpose(-2,-1) / math.sqrt(query.shape[-1])
```

Comment:
you can matmul `query` and `key` with `acc_dtype=dtypes.float32` to get float32 from matmul, I think we want to patch the core version instead of adding a separate one.

Reply:
also not sure what triggered the nan, even if there are `-inf` into softmax, as long as there's one finite number it should be fine.

Reply:
@chenyuxyz Yes, but the infs are one full row that we softmax over.

Reply:
Ok, so I have figured out what is going on. The `inf`s are a result of under and overflow which itself is not an issue. However, if there are enough `inf` to fill a full row over the softmax dimension after that, we get `NaN`. That means that `acc_dtype=dtypes.float32` wont fix it unless we add some flag that allows us to do the matmul without casting to lower precision right after the accumulation.

Reply:
oh i can make dot return in acc_dtype if it's specified

Reply:
added in #4994, now if acc_dtype is explicitly set, it does not downcast